### ISSUE TITLE
[Store] Getting relationship data

### ISSUE THREAD

**Issue by Jopie01** [emberjs/data#8553]
I'm struggling with getting data out of relational fields. I have a few `@ember-data` packages installed:
```js
  "@ember-data/graph": "4.12.0-alpha.20",
  "@ember-data/json-api": "4.12.0-alpha.20",
  "@ember-data/request": "4.12.0-alpha.20",
  "@ember-data/store": "4.12.0-alpha.20",
  "@ember-data/tracking": "4.12.0-alpha.20",
```
I created my own `SchemaDefinitionService` and I'm able to push JSONAPI data into the store. Debugging shows that everything went well even the relational schemas are created. When getting the data through the `instantiateRecord` hook, the `attributes` and `relationships` keys are there. However for the relationships I'm missing the `attribute` key, only the `lid` and `id` etc are there.

Should I get those data manually running through a for-loop in the `instantiateRecord`? The same as I do for the 'main' record?

**Comment by runspired** [comment]
Right, we return the identifiers not the resource data they point to.

You would want to iterate the relationship data calling `store.peekRecord` for any resources you expect to be present in the cache, or if you want it all as part of one thing you can fetch the data for just that resource from the cache and add it into your single object.

Broad strokes: there's a missing API here that needs RFC'd for `Store.peek` which would take a document identifier and return the hydrated Document like `Store.request` gives you. Relationships are (from a meta perspective) usually pointers between documents.

**Comment by Jopie01** [comment]
Hmm, I get two different results. I have a nested menu and I fetch the parts in different requests:

1. The top level is fetched with its related menu items (minimum data)
2. The stored tree structure (which menu items are expanded) is fetched
3. Based on the structure relationships are enriched with more data

The requests are all the same but I get different results (this is in the `store` service):

- using `this.cache.peek(identifier)` I get the record back with the enriched data from 3.
- using `this.peekRecord(identifier)` I get the minimum back from 1.

When on a menu item is clicked, the underlying menu items will then be fetched from the backend. The data returned has the same behaviour as described.

Is this expected behaviour?

**Comment by runspired** [comment]
It sounds expected, it just depends on what you've implemented `instantiateRecord` to do. `peekRecord` will ask the cache if it has the data for that resource, then hand you the record instance associated to it, instantiating it via that hook if none exists yet.

`cache.peek` will give you whatever you've configured the cache to hand you as the resource data, which by default is a JSON:API resource with relationships that contain identifiers to other resources.

The intelligent thing to do here is to hand back a Proxy that lazily fetches more of the graph when accessed. E.g. when the relationship `foos` is accessed it would iterate the resource data calling `peekRecord` on each contained identity.

**Comment by Jopie01** [comment]
I have to come back to this. I had a big fight with getting the data from cache. In the end it turned out that it was my fault that it didn't work because of weird things in the `instantiateRecord` hook. Have to first clean things up now.

**Comment by Jopie01** [comment]
Indeed I messed things up :cry: Sorry for wasting your time.

I want to have both `attributes` and `relationships` on one level.

```js
// current:
{
  id: "123",
  lid: "@lid:menu-123",
  type: "menu",
  attributes: {
    name: "test menu",
    level: 2,
    id: 123,
    icon: "settings" 
  },
  relationships: {
    user_created: {
      id: 0,
      name: "root"
    }
  }
}

// I want:
{
  id: "123",
  lid: "@lid:menu-123",
  type: "menu",
  name: "test menu",
  level: 2,
  id: 123,
  icon: "settings",
  user_created: {
    id: 0,
    name: "root"
  }
}
```

I had the `instantiateRecord` hook from the example in the `store` README but instead of returning the whole TrackedObject I returned parts of it basically dropping the reactivity. Also the notification got through but the data got mangled.

Also I wanted to replace the record in the route which I should not do. Just call the `this.store.findRecord()` in my case and the record magically updates with the right fields and data!

As I now have almost cleaned things up (still have to de-duplicate code, everything works beautifully!
